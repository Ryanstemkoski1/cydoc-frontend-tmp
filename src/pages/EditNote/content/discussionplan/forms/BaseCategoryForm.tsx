/**
 * @fileoverview Base components and types to abstract the overall
 * structure/responsiveness/interactions of the discussion and plan page.
 */

import {
    DiagnosesOptionMapping,
    OptionMapping,
    getOptionMapping,
} from '_processOptions';
import AddRowButton from 'components/tools/AddRowButton/AddRowButton';
import { WhenResponse } from 'constants/enums';
import medications from 'constants/medications';
import procedures from 'constants/procedures';
import registrationConstants from 'constants/registration-constants.json';
import _ from 'lodash';
import React, { useEffect, useState } from 'react';
import { Accordion, Grid, Header, Icon } from 'semantic-ui-react';
import {
    ConditionCategoryKey,
    EventHandler,
    HandleOnAddItem,
    PlanAction,
} from '../util';

type Options = { main: OptionMapping; when?: OptionMapping };

type DiagnosesOptions = {
    main: DiagnosesOptionMapping;
    when?: DiagnosesOptionMapping;
};

interface CategoryFormStateProps<T> {
    categoryData: T[];
}

export interface CategoryFormOwnProps {
    mobile: boolean;
    conditionId: string;
    formatAction: (action: PlanAction, ...args: any[]) => EventHandler;
}

export type CategoryFormProps<T> = CategoryFormOwnProps &
    CategoryFormStateProps<T>;

export type CategoryFormComponent<T> = (
    row: T,
    options?: Options | DiagnosesOptions,
    onAddItem?: HandleOnAddItem
) => JSX.Element;

interface BaseCategoryFormProps<T> {
    mobile: boolean;
    category: ConditionCategoryKey;
    categoryData: T[];
    numColumns: 2 | 3;
    addRowLabel: string;
    components: {
        mobileContent: CategoryFormComponent<T>;
        mobileTitle: CategoryFormComponent<T>;
        gridColumn: CategoryFormComponent<T>;
        gridHeaders: React.FC;
    };
    addRow: EventHandler;
}

const specialties = getOptionMapping(registrationConstants.specialties);

const TYPE_TO_OPTION: {
    [key in ConditionCategoryKey]: OptionMapping;
} = {
    differentialDiagnoses: {},
    prescriptions: medications,
    proceduresAndServices: procedures,
    referrals: specialties,
};

/**
 * Component for all categories of the discussion and plan section of the note
 */
export const BaseCategoryForm = <T extends { id: string }>(
    props: BaseCategoryFormProps<T>
) => {
    const {
        category,
        categoryData,
        mobile,
        components,
        numColumns,
        addRowLabel,
        addRow,
    } = props;

    const [expanded, setExpanded] = useState<boolean>(false);
    const [activeRows, setActiveRows] = useState(new Set());
    const [options, setOptions] = useState<Options | DiagnosesOptions>({
        main: TYPE_TO_OPTION[category] || {},
    });

    useEffect(() => {
        if (category === 'differentialDiagnoses') {
            import('constants/diagnoses.json').then((diseases) => {
                setOptions({
                    main: diseases as unknown as DiagnosesOptionMapping,
                });
            });
        }
    }, [category]);

    useEffect(() => {
        if (categoryData.length > 0 && 'when' in categoryData[0]) {
            const whenOptions = Object.values(WhenResponse)
                .filter((res) => res !== WhenResponse.None)
                .reduce((options, res) => {
                    const label = res.replace('_', ' ').toLowerCase();
                    options[res] = { label, value: res };
                    return options;
                }, {} as OptionMapping);
            setOptions({
                ...options,
                when: whenOptions,
            });
        }
    }, [categoryData]);

    const onAddItem: HandleOnAddItem = (_e, { optiontype, value }) => {
        // Add to list of options
        setOptions({
            ...options,
            [optiontype]: {
                ...options[optiontype as 'main' | 'when'],
                [value]: { value, label: value },
            },
        });
    };

    // Render each entry as an expandable accordion when in mobile view, and as a
    // responsive grid otherwise.
    let content;
    if (mobile) {
        const toggleRow = (id: string) => {
            const newActiveRows = new Set(activeRows);
            if (activeRows.has(id)) {
                newActiveRows.delete(id);
            } else {
                newActiveRows.add(id);
            }
            setActiveRows(newActiveRows);
        };

        const panels = categoryData.map((row) => {
            const title = components.mobileTitle(row, options, onAddItem);
            const content = components.mobileContent(row, options, onAddItem);
            return {
                key: row.id,
                active: activeRows.has(row.id),
                title: { content: title },
                onTitleClick: () => toggleRow(row.id),
                content: { content },
            };
        });

        content = (
            <Accordion
                fluid
                styled
                exclusive={false}
                panels={panels}
                className={`plan-section-form ${category as string}`}
            />
        );
    } else {
        const body = categoryData.map((row) =>
            components.gridColumn(row, options, onAddItem)
        );
        content = (
            <Grid stackable columns={numColumns} className='section-body'>
                {components.gridHeaders({})}
                {body}
            </Grid>
        );
    }

    const checkDifferentialDiagnoses = (s: string) => {
        return s == 'differentialDiagnoses'
            ? 'Diagnosis'
            : _.startCase(s as string);
    };

    // const enableToggle = mobile && category !== 'differentialDiagnoses';
    const enableToggle = mobile;

    return (
        <Accordion fluid className='plan-section'>
            <Accordion.Title
                active={!enableToggle || expanded}
                /* eslint-disable-next-line */
                onClick={enableToggle ? () => setExpanded(!expanded) : () => {}}
            >
                <Header as='h2' attached size='large'>
                    {mobile &&
                        (expanded ? (
                            <Icon name='caret down'></Icon>
                        ) : (
                            <Icon name='caret right'></Icon>
                        ))}

                    {_.startCase(category as string)}
                </Header>
            </Accordion.Title>
            <Accordion.Content active={!enableToggle || expanded}>
                {content}
                <AddRowButton name={addRowLabel} onClick={addRow} />
            </Accordion.Content>
        </Accordion>
    );
};
