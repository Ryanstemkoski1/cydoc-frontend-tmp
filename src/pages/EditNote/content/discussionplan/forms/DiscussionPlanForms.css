.dropdown__placeholder {
    opacity: 0.3;
}
