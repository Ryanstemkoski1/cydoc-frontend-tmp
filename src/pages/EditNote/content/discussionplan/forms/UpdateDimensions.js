import { useState, useEffect } from 'react';

export default function UpdateDimensions() {
    function getDimensions() {
        const width = typeof window !== 'undefined' ? window.innerWidth : 0;
        return { width };
    }

    const [windowDimensions, setWindowDimensions] = useState(getDimensions());

    function handleResize() {
        setWindowDimensions(getDimensions());
    }

    useEffect(() => {
        if (typeof window !== 'undefined') {
            window.addEventListener('resize', handleResize);
            return () => window.removeEventListener('resize', handleResize);
        }
    }, [typeof window !== 'undefined', handleResize]);

    return windowDimensions;
}
