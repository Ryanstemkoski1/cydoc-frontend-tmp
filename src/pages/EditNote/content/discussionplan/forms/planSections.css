/* styling for DifferentialDiagnosesForm */
.mobile-title {
    margin-top: 10px;
    font-weight: normal;
}

#mobile-content-form {
    width: 95%;
}

#main-input-div {
    width: 270px;
    height: 15px;
}

/* mobile screen */
@media (max-width: 450px) {
    #container-div {
        height: 5px;
    }
}

@media (max-width: 380px) {
    #main-input-div {
        width: 180px;
        height: 15px;
    }
}

/* styling for PrescriptionsForm */
.prescriptions-label {
    margin-left: 15px;
}

#prescriptions-main-input-container {
    margin-top: 0px;
    margin-bottom: 15px;
}

#prescriptions-container1 {
    height: 5px;
}

#prescriptions-container2 {
    height: 10px;
}

#prescriptions-container3 {
    height: 20px;
}

#prescriptions-ui-form {
    width: 90%;
    margin: auto;
}

/* styling for MainWhenComments */
#main-when-mobile-title {
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
}

#near-width {
    width: 95%;
}

/* smaller/mobile screens */
@media (max-width: 800px) {
    /* MainWhenComments */
    #main-when-mobile-title {
        width: 85%;
        margin: auto;
        margin-top: 10px;
        margin-bottom: 15px;
    }

    #main-when-mobile-title-inner {
        height: 5px;
    }

    #main-when-mobile-content {
        height: 20px;
    }

    /* prescriptions */
    #prescriptions-main-input-container {
        width: 90%;
        margin-top: 10px;
        margin-bottom: 15px;
    }
}