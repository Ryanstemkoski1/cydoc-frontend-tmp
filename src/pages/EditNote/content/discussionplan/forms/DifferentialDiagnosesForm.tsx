import Dropdown from 'components/tools/RecursiveDropdown';
import React from 'react';
import { connect } from 'react-redux';
import {
    addDifferentialDiagnosis,
    updateDifferentialDiagnosis,
    updateDifferentialDiagnosisComments,
} from '@redux/actions/planActions';
import { CurrentNoteState } from '@redux/reducers';
import {
    PlanDiagnosisFlat,
    selectPlanCondition,
} from '@redux/selectors/planSelectors';
import { Grid, TextArea } from 'semantic-ui-react';
import { PlanAction } from '../util';
import {
    BaseCategoryForm,
    CategoryFormComponent,
    CategoryFormOwnProps,
    CategoryFormProps,
} from './BaseCategoryForm';
// import UpdateDimensions from './UpdateDimensions';
import { DiagnosesOptionMapping } from '_processOptions';
import './DiscussionPlanForms.css';
import './planSections.css';

interface DifferentialDiagnosesDispatchProps {
    addDifferentialDiagnosis: PlanAction;
    updateDifferentialDiagnosisComments: PlanAction;
    updateDifferentialDiagnosis: PlanAction;
}

type ComponentFunction = CategoryFormComponent<PlanDiagnosisFlat>;

const DifferentialDiagnosesForm = (
    props: CategoryFormProps<PlanDiagnosisFlat> &
        DifferentialDiagnosesDispatchProps
) => {
    const { categoryData, formatAction, conditionId, ...actions } = props;

    const gridHeaders = () => (
        <Grid.Row>
            <Grid.Column>Diagnosis</Grid.Column>
            <Grid.Column>Comments</Grid.Column>
        </Grid.Row>
    );

    const mainInput: ComponentFunction = (row, options) => (
        <>
            <Dropdown
                fluid
                search
                selection
                clearable
                loading={options && Object.keys(options.main).length == 0}
                disabled={options && Object.keys(options.main).length == 0}
                transparent={false}
                value={row.diagnosis}
                code={row.code}
                options={(options?.main as DiagnosesOptionMapping) || {}}
                uuid={row.id}
                onChange={formatAction(actions.updateDifferentialDiagnosis)}
                optiontype='main'
                aria-label='Diagnosis-Dropdown'
                placeholder='diagnosis'
                className='main-input'
            />
            <div className='container' id='main-input-div' />
        </>
    );

    const gridColumn: ComponentFunction = (row, options, onAddItem) => {
        return (
            <React.Fragment key={row.id}>
                <Grid.Column width={8}>
                    {mainInput(row, options, onAddItem)}
                </Grid.Column>
                <Grid.Column>
                    <div className='ui form'>
                        <TextArea
                            uuid={row.id}
                            onChange={formatAction(
                                actions.updateDifferentialDiagnosisComments
                            )}
                            value={row.comments}
                            aria-label='Diagnosis-Comment'
                            placeholder='e.g. this diagnosis is more likely because...'
                        />
                    </div>
                </Grid.Column>
            </React.Fragment>
        );
    };

    return (
        <BaseCategoryForm<PlanDiagnosisFlat>
            numColumns={2}
            addRowLabel='diagnosis'
            category='differentialDiagnoses'
            categoryData={categoryData}
            addRow={() => actions.addDifferentialDiagnosis(conditionId)}
            // addRow={formatAction(actions.addDifferentialDiagnosis)}
            components={{
                gridColumn,
                gridHeaders,
                /* eslint-disable-next-line */
                // gridHeaders: () => <></>,
            }}
        />
    );
};

export default connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .differentialDiagnoses,
    }),
    {
        addDifferentialDiagnosis,
        updateDifferentialDiagnosisComments,
        updateDifferentialDiagnosis,
    }
)(DifferentialDiagnosesForm);
