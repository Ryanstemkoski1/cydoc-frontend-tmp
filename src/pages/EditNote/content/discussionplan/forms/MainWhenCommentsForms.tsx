/**
 * @fileoverview Components that utilizes [main: string], when, comments fields
 * in the discussion and plan page
 */
import { OptionMapping } from '_processOptions';
import Dropdown from 'components/tools/OptimizedDropdown';
import { WhenResponse } from 'constants/enums';
import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import {
    addProcedureOrService,
    addReferral,
    updateProcedureOrService,
    updateProcedureOrServiceComments,
    updateProcedureOrServiceWhen,
    updateReferralComments,
    updateReferralDepartment,
    updateReferralWhen,
} from 'redux/actions/planActions';
import { CurrentNoteState } from 'redux/reducers';
import {
    PlanProceduresAndServicesFlat,
    PlanReferralsFlat,
    selectPlanCondition,
} from 'redux/selectors/planSelectors';
import { Grid, TextArea } from 'semantic-ui-react';
import { ConditionCategoryKey, PlanAction } from '../util';
import {
    BaseCategoryForm,
    CategoryFormComponent,
    CategoryFormOwnProps,
    CategoryFormProps,
} from './BaseCategoryForm';
import './DiscussionPlanForms.css';
import './planSections.css';

export const ProceduresAndServicesForm = connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .proceduresAndServices,
    }),
    {
        mainOnChange: updateProcedureOrService,
        whenOnChange: updateProcedureOrServiceWhen,
        commentsOnChange: updateProcedureOrServiceComments,
        addRow: addProcedureOrService,
    }
)(
    (
        props: CategoryFormProps<PlanProceduresAndServicesFlat> &
            MainWhenCommentsDispatchProps
    ) => (
        <MainWhenCommentsForm<PlanProceduresAndServicesFlat>
            {...props}
            categoryProps={{
                category: 'proceduresAndServices',
                mainValueName: 'procedure',
                addRowLabel: 'procedure or service',
                getMainValue: (row) => row.procedure,
            }}
        />
    )
);

export const ReferralsForm = connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .referrals,
    }),
    {
        mainOnChange: updateReferralDepartment,
        whenOnChange: updateReferralWhen,
        commentsOnChange: updateReferralComments,
        addRow: addReferral,
    }
)(
    (
        props: CategoryFormProps<PlanReferralsFlat> &
            MainWhenCommentsDispatchProps
    ) => (
        <MainWhenCommentsForm<PlanReferralsFlat>
            {...props}
            categoryProps={{
                category: 'referrals',
                mainValueName: 'department',
                addRowLabel: 'department',
                getMainValue: (row) => row.department,
            }}
        />
    )
);

interface MainWhenCommentsDispatchProps {
    addRow: PlanAction;
    mainOnChange: PlanAction;
    whenOnChange: PlanAction;
    commentsOnChange: PlanAction;
}

type MainWhenCommentsOwnProps<T> = CategoryFormProps<T> & {
    categoryProps: {
        category: ConditionCategoryKey;
        mainValueName:
            | keyof PlanProceduresAndServicesFlat
            | keyof PlanReferralsFlat;
        addRowLabel: string;
        getMainValue: (row: T) => string;
    };
};

type MainWhenCommentsFormProps<T> = MainWhenCommentsDispatchProps &
    MainWhenCommentsOwnProps<T>;

/**
 * Base component for forms with {[main: string], when, comments} fields
 */
export const MainWhenCommentsForm = <
    T extends { id: string; when: WhenResponse; comments: string },
>(
    props: MainWhenCommentsFormProps<T>
) => {
    const { categoryData, categoryProps, formatAction, ...actions } = props;

    const gridHeaders = () => (
        <Grid.Row>
            <Grid.Column>
                {_.upperFirst(categoryProps.mainValueName as string)}{' '}
            </Grid.Column>
            <Grid.Column> When </Grid.Column>
            <Grid.Column> Comments </Grid.Column>
        </Grid.Row>
    );

    const mobileTitle: CategoryFormComponent<T> = (row, options, onAddItem) => (
        <div className='container' id='main-when-mobile-title'>
            <div className='container' id='main-when-mobile-title-inner' />
            <Dropdown
                fluid
                search
                selection
                allowAdditions
                clearable
                transparent={false}
                optiontype='main'
                uuid={row.id}
                options={(options?.main as OptionMapping) || {}}
                onChange={formatAction(actions.mainOnChange)}
                onAddItem={onAddItem}
                value={categoryProps.getMainValue(row)}
                placeholder={categoryProps.mainValueName}
                aria-label={`${categoryProps.mainValueName}-Dropdown`}
                className='main-input'
            />
        </div>
    );

    const whenInput: CategoryFormComponent<T> = (row, options, onAddItem) => (
        <>
            <div className='container' id='main-when-mobile-title-inner' />
            <div id='near-width'>
                <Dropdown
                    fluid
                    search
                    selection
                    allowAdditions
                    clearable
                    transparent={false}
                    optiontype='when'
                    uuid={row.id}
                    options={(options?.when as OptionMapping) || {}}
                    onChange={formatAction(actions.whenOnChange)}
                    onAddItem={onAddItem}
                    value={row.when}
                    aria-label={`${categoryProps.mainValueName}-When`}
                    placeholder='when'
                    className='expanded-input'
                />
            </div>
        </>
    );

    const gridColumn: CategoryFormComponent<T> = (row, options, onAddItem) => (
        <React.Fragment key={row.id}>
            <Grid.Column>{mobileTitle(row, options, onAddItem)}</Grid.Column>
            <Grid.Column>{whenInput(row, options, onAddItem)}</Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        name='comment'
                        uuid={row.id}
                        value={row.comments}
                        onChange={formatAction(actions.commentsOnChange)}
                        aria-label={`${
                            categoryProps.mainValueName as string
                        }-Comment`}
                        placeholder='comments'
                    />
                </div>
            </Grid.Column>
        </React.Fragment>
    );

    return (
        <BaseCategoryForm
            category={categoryProps.category}
            categoryData={categoryData}
            numColumns={3}
            addRowLabel={categoryProps.addRowLabel}
            // addRow={formatAction(actions.addRow)}
            addRow={() => actions.addRow(actions.conditionId)}
            components={{
                gridColumn,
                gridHeaders,
            }}
        />
    );
};
