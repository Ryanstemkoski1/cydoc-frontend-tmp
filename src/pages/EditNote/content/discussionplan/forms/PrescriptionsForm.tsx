import React from 'react';
import { Grid, TextArea } from 'semantic-ui-react';
import Dropdown from 'components/tools/OptimizedDropdown';
import { PlanAction } from '../util';
import { connect } from 'react-redux';
import { CurrentNoteState } from 'redux/reducers';
import {
    PlanPrescriptionFlat,
    selectPlanCondition,
} from 'redux/selectors/planSelectors';
import {
    addPrescription,
    updatePrescriptionComments,
    updatePrescriptionDose,
    updatePrescriptionType,
    updatePrescriptionSignature,
} from 'redux/actions/planActions';
import {
    CategoryFormProps,
    CategoryFormComponent,
    CategoryFormOwnProps,
    BaseCategoryForm,
} from './BaseCategoryForm';
import UpdateDimensions from './UpdateDimensions';
import './planSections.css';

interface PrescriptionsDispatchProps {
    addPrescription: PlanAction;
    updatePrescriptionComments: PlanAction;
    updatePrescriptionDose: PlanAction;
    updatePrescriptionType: PlanAction;
    updatePrescriptionSignature: PlanAction;
}

type ComponentFunction = CategoryFormComponent<PlanPrescriptionFlat>;

const PrescriptionsForm = (
    props: CategoryFormProps<PlanPrescriptionFlat> & PrescriptionsDispatchProps
) => {
    const { mobile, categoryData, formatAction, ...actions } = props;
    const { width } = UpdateDimensions();

    const gridHeaders = () => (
        <Grid.Row>
            <Grid.Column> Rx </Grid.Column>
            <Grid.Column> Signature (Sig) </Grid.Column>
            <Grid.Column> Comments </Grid.Column>
        </Grid.Row>
    );

    const mainInput: ComponentFunction = (row, options, onAddItem) => (
        <div className='container' id='prescriptions-main-input-container'>
            <Dropdown
                fluid
                search
                selection
                allowAdditions
                clearable
                optiontype='main'
                transparent={mobile}
                uuid={row.id}
                value={row.type}
                options={options?.main || {}}
                onChange={formatAction(actions.updatePrescriptionType)}
                onAddItem={onAddItem}
                aria-label='Prescription-Dropdown'
                placeholder='medication name'
                className='main-input recipe'
            />
            <div className='container' id='prescriptions-container3' />
            <label>Prescription Dose</label>
            <div className='container' id='prescriptions-container2' />
            <div className='ui form'>
                <TextArea
                    fluid
                    type='text'
                    transparent={mobile}
                    uuid={row.id}
                    value={row.dose}
                    onChange={formatAction(actions.updatePrescriptionDose)}
                    aria-label='Prescription-Amount'
                    placeholder='e.g. 81 mg tablet'
                    className={`recipe-amount ${!mobile && 'lg'}`}
                />
            </div>
        </div>
    );

    const gridColumn: ComponentFunction = (row, options, onAddItem) => (
        <React.Fragment key={row.id}>
            <Grid.Column>{mainInput(row, options, onAddItem)}</Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionSignature
                        )}
                        value={row.signature}
                        aria-label='Prescription-Signature'
                        placeholder='e.g. 1 tablet every 8 hours'
                    />
                </div>
            </Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionComments
                        )}
                        value={row.comments}
                        aria-label='Prescription-Comment'
                        placeholder='e.g. take with food'
                    />
                </div>
            </Grid.Column>
        </React.Fragment>
    );

    const mobileTitle: ComponentFunction = (row, options, onAddItem) => (
        <div className='recipe'>
            <label> Rx </label>
            {mainInput(row, options, onAddItem)}
        </div>
    );

    const mobileContent: ComponentFunction = (row) => (
        <>
            <label className='prescriptions-label'> Signature (Sig) </label>
            <div className='container' id='prescriptions-container1' />
            <div className='ui form' id='prescriptions-ui-form'>
                <TextArea
                    fluid
                    transparent
                    type='text'
                    uuid={row.id}
                    value={row.signature}
                    onChange={formatAction(actions.updatePrescriptionSignature)}
                    placeholder='e.g. 1 tablet every 8 hours'
                    aria-label='Prescription-Signature'
                    className='expanded-input'
                />
            </div>
            <div className='container' id='prescriptions-container2' />
            <label className='prescriptions-label'> Comments </label>
            <div className='container' id='prescriptions-container1' />
            <div className='ui form' id='prescriptions-ui-form'>
                <TextArea
                    fluid
                    transparent
                    type='text'
                    uuid={row.id}
                    value={row.comments}
                    onChange={formatAction(actions.updatePrescriptionComments)}
                    placeholder='e.g. take with food'
                    aria-label='Prescription-Comment'
                    className='expanded-input'
                />
            </div>
        </>
    );

    return (
        <BaseCategoryForm<PlanPrescriptionFlat>
            mobile={mobile}
            category='prescriptions'
            categoryData={categoryData}
            numColumns={3}
            addRowLabel='prescription'
            addRow={formatAction(actions.addPrescription)}
            components={{
                gridColumn,
                gridHeaders,
                mobileTitle,
                mobileContent,
            }}
        />
    );
};

export default connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .prescriptions,
    }),
    {
        addPrescription,
        updatePrescriptionComments,
        updatePrescriptionDose,
        updatePrescriptionType,
        updatePrescriptionSignature,
    }
)(PrescriptionsForm);
