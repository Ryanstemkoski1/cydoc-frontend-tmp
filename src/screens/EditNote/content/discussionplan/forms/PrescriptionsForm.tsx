import { OptionMapping } from '../../../../../_processOptions';
import Dropdown from '@components/tools/OptimizedDropdown';
import React from 'react';
import { connect } from 'react-redux';
import {
    addPrescription,
    updatePrescriptionComments,
    updatePrescriptionDose,
    updatePrescriptionSignature,
    updatePrescriptionType,
} from '@redux/actions/planActions';
import { CurrentNoteState } from '@redux/reducers';
import {
    PlanPrescriptionFlat,
    selectPlanCondition,
} from '@redux/selectors/planSelectors';
import { Grid, TextArea } from 'semantic-ui-react';
import { PlanAction } from '../util';
import {
    BaseCategoryForm,
    CategoryFormComponent,
    CategoryFormOwnProps,
    CategoryFormProps,
} from './BaseCategoryForm';
import './DiscussionPlanForms.css';
import useUpdateDimensions from './useUpdateDimensions';
import './planSections.css';

interface PrescriptionsDispatchProps {
    addPrescription: PlanAction;
    updatePrescriptionComments: PlanAction;
    updatePrescriptionDose: PlanAction;
    updatePrescriptionType: PlanAction;
    updatePrescriptionSignature: PlanAction;
}

type ComponentFunction = CategoryFormComponent<PlanPrescriptionFlat>;

const PrescriptionsForm = (
    props: CategoryFormProps<PlanPrescriptionFlat> & PrescriptionsDispatchProps
) => {
    const { categoryData, formatAction, ...actions } = props;
    useUpdateDimensions();

    const gridHeaders = () => (
        <Grid.Row>
            <Grid.Column> Rx </Grid.Column>
            <Grid.Column> Signature (Sig) </Grid.Column>
            <Grid.Column> Comments </Grid.Column>
        </Grid.Row>
    );

    const mainInput: ComponentFunction = (row, options, onAddItem) => (
        <div className='container' id='prescriptions-main-input-container'>
            <Dropdown
                fluid
                search
                selection
                allowAdditions
                clearable
                optiontype='main'
                transparent={false}
                uuid={row.id}
                value={row.type}
                options={(options?.main as OptionMapping) || {}}
                onChange={formatAction(actions.updatePrescriptionType)}
                onAddItem={onAddItem}
                aria-label='Prescription-Dropdown'
                placeholder='medication name'
                className='main-input recipe'
            />
            <div className='container' id='prescriptions-container3' />
            <label>Prescription Dose</label>
            <div className='container' id='prescriptions-container2' />
            <div className='ui form'>
                <TextArea
                    type='text'
                    uuid={row.id}
                    value={row.dose}
                    onChange={formatAction(actions.updatePrescriptionDose)}
                    aria-label='Prescription-Amount'
                    placeholder='e.g. 81 mg tablet'
                    className={`recipe-amount lg`}
                />
            </div>
        </div>
    );

    const gridColumn: ComponentFunction = (row, options, onAddItem) => (
        <React.Fragment key={row.id}>
            <Grid.Column>{mainInput(row, options, onAddItem)}</Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionSignature
                        )}
                        value={row.signature}
                        aria-label='Prescription-Signature'
                        placeholder='e.g. 1 tablet every 8 hours'
                    />
                </div>
            </Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionComments
                        )}
                        value={row.comments}
                        aria-label='Prescription-Comment'
                        placeholder='e.g. take with food'
                    />
                </div>
            </Grid.Column>
        </React.Fragment>
    );

    return (
        <BaseCategoryForm<PlanPrescriptionFlat>
            category='prescriptions'
            categoryData={categoryData}
            numColumns={3}
            addRowLabel='prescription'
            // addRow={formatAction(actions.addPrescription)}
            addRow={() => actions.addPrescription(actions.conditionId)}
            components={{
                gridColumn,
                gridHeaders,
            }}
        />
    );
};

export default connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .prescriptions,
    }),
    {
        addPrescription,
        updatePrescriptionComments,
        updatePrescriptionDose,
        updatePrescriptionType,
        updatePrescriptionSignature,
    }
)(PrescriptionsForm);
