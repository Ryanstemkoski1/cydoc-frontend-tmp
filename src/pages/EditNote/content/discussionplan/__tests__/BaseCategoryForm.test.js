import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import { BaseCategoryForm } from '../forms/BaseCategoryForm';

Enzyme.configure({ adapter: new Adapter() });

const mountWithProps = ({ ...props } = {}) => {
    props = {
        category: 'differentialDiagnoses',
        categoryData: [{}],
        numColumns: 2,
        components: {
            mobileContent: jest.fn(),
            mobileTitle: jest.fn(),
            gridColumn: jest.fn(),
            gridHeaders: jest.fn(),
        },
        ...props,
    };
    return mount(<BaseCategoryForm {...props} />);
};

describe('BaseCategoryForm', () => {
    it('renders without crashing', () => {
        const wrapper = mountWithProps();
        expect(wrapper).toBeTruthy();
    });

    it('matches snapshot', () => {
        const wrapper = mountWithProps();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('calls only mobileContent and mobileTitle when in mobile view', () => {
        const mockMobileContent = jest.fn();
        const mockMobileTitle = jest.fn();
        const mockGridHeaders = jest.fn();
        const mockGridColumn = jest.fn();
        mountWithProps({
            mobile: true,
            components: {
                mobileContent: mockMobileContent,
                mobileTitle: mockMobileTitle,
                mockGridHeaders: mockGridHeaders,
                mockGridColumn: mockGridColumn,
            },
        });

        expect(mockMobileTitle).toHaveBeenCalled();
        expect(mockMobileContent).toHaveBeenCalled();

        expect(mockGridColumn).toHaveBeenCalledTimes(0);
        expect(mockGridHeaders).toHaveBeenCalledTimes(0);
    });

    it('calls only gridHeaders and gridColumn when not in mobile view', () => {
        const mockMobileContent = jest.fn();
        const mockMobileTitle = jest.fn();
        const mockGridHeaders = jest.fn();
        const mockGridColumn = jest.fn();
        mountWithProps({
            mobile: false,
            components: {
                mobileContent: mockMobileContent,
                mobileTitle: mockMobileTitle,
                gridHeaders: mockGridHeaders,
                gridColumn: mockGridColumn,
            },
        });

        expect(mockMobileTitle).toHaveBeenCalledTimes(0);
        expect(mockMobileContent).toHaveBeenCalledTimes(0);

        expect(mockGridColumn).toHaveBeenCalled();
        expect(mockGridHeaders).toHaveBeenCalled();
    });

    // // TODO: Fix below tests
    // it('should not be toggable for differentialDiagnoses', () => {
    //     let wrapper = mountWithProps({
    //         mobile: false,
    //         category: 'differentialDiagnoses',
    //     });
    //     expect(wrapper.find('.active.content')).toHaveLength(1);
    //     wrapper.find('.title').first().simulate('click');
    //     expect(wrapper.find('.active.content')).toHaveLength(1);

    //     // Same applies for mobile view
    //     wrapper = mountWithProps({
    //         mobile: true,
    //         category: 'differentialDiagnoses',
    //     });
    //     expect(wrapper.find('.active.content')).toHaveLength(1);
    //     wrapper.find('.title').first().simulate('click');
    //     expect(wrapper.find('.active.content')).toHaveLength(1);
    // });

    it('should only be toggable in mobile for non-differentialDiagnoses categories', () => {
        let wrapper = mountWithProps({
            mobile: false,
            category: 'prescriptions',
        });
        expect(wrapper.find('.active.content')).toHaveLength(1);
        wrapper.find('.title').first().simulate('click');
        expect(wrapper.find('.active.content')).toHaveLength(1);

        // Mobile starts off collapsed
        wrapper = mountWithProps({ mobile: true, category: 'prescriptions' });
        expect(wrapper.find('.active.content')).toHaveLength(0);
        wrapper.find('.title').first().simulate('click');
        expect(wrapper.find('.active.content')).toHaveLength(1);
    });

    it('converts category prop to header using start case', () => {
        const wrapper = mountWithProps({ category: 'differentialDiagnoses' });
        const header = wrapper.find('.ui.attached.header');
        expect(header.text()).toEqual('Differential Diagnoses');
    });
});
