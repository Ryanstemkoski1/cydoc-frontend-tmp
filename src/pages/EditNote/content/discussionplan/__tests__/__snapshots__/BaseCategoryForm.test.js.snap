// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCategoryForm matches snapshot 1`] = `"<div class=\\"accordion ui fluid plan-section\\"><div class=\\"active title\\"><h2 class=\\"ui large attached header\\">Differential Diagnoses</h2></div><div class=\\"content active\\"><div class=\\"ui stackable two column grid section-body\\"></div><button class=\\"addButton\\"><img src=\\"add.svg\\" alt=\\"Add\\"><span>Add </span></button></div></div>"`;
